### PR TITLE
[GPU] Implement GridSample-9

### PR DESCRIPTION
**Pull request by mhnap-lohika**
### Details:
 - Added GPU implementation
 - Added single-layer tests
 - Added unit tests
 - Added blocked layout support

### Tickets:
 - 88015

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/grid_sample/grid_sample_kernel_ref.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    return key;
+}
+
+bool GridSampleKernelRef::Validate(const Params& params, const optional_params& options) const {
+    if (params.GetType() != KernelType::GRID_SAMPLE || options.GetType() != KernelType::GRID_SAMPLE) {
+        return false;
+    }
+
+    const auto& kernel_params = dynamic_cast<const grid_sample_params&>(params);
+    if (kernel_params.inputs.size() != 2) {
```
Comment (kelvinchoi-intel) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/grid_sample/grid_sample_kernel_ref.cpp:62]: The two inputs should have 4-dim tensor. Why don't you check this at this point?

**Comment by mhnap-lohika** [comment]
@wilson-seok Seems that GridSample op supports only float precision. We have such code in src/core/src/op/grid_sample.cpp:
`NGRAPH_TYPE_CASE(evaluate_grid_sample, f32, output, data, grid, attributes);`
But I cannot change the already existing implementation, so there are no single-layer tests for other precision than float.

**Comment by mhnap-lohika** [comment]
> > grid_sample.cpp
> 
> @mhnap-lohika Does other precision such as int32 cause evaluate error in the op evaluate process? Op reference test seems to use only f32 for data precision also.https://github.com/openvinotoolkit/openvino/blob/9ea33715a082ae23f085898f7bf08ea973b57f00/src/plugins/template/tests/functional/op_reference/grid_sample.cpp And if this PR is ready for review, please switch to 'ready for review'.

Yes, any integer type causes an evaluation error. So we test only float types.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl [hunk 2]
```diff
@@ -1,40 +1,240 @@
-// Copyright (C) 2022 Intel Corporation
```
Comment (mhnap-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl]: @ozhydkov-lohika Please add back the copyright.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl [hunk 3]
```diff
@@ -1,40 +1,240 @@
...
+    return v;
+}
+#define cubic_coeffs FUNC_CALL(cubic_coeffs)
+
+inline matrix_data_4x4_t FUNC(
+    gather_4x4)(const data_t* data, const size_t n, const size_t c, const long y_topleft, const long x_topleft) {
+    matrix_data_4x4_t s;
+    for (int j = 0; j < 4; ++j)
+        for (int i = 0; i < 4; ++i)
+            s[j * 4 + i] = get_padded(data, n, c, y_topleft + j, x_topleft + i);
```
Comment (mhnap-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl:161]: @ozhydkov-lohika  I think it's better to use curly braces in condition statements.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl [hunk 4]
```diff
@@ -1,40 +1,240 @@
...
+typedef __global INPUT0_TYPE data_t;
+typedef __global INPUT1_TYPE grid_t;
+typedef __global OUTPUT_TYPE output_t;
+
+typedef INPUT0_TYPE data_et;
+typedef INPUT1_TYPE grid_et;
+typedef OUTPUT_TYPE output_et;
+
+inline const data_et FUNC(
+    get_data_single_value)(const data_t* buffer, const size_t n, const size_t c, const size_t h, const size_t w) {
```
Comment (mhnap-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl:10]: @ozhydkov-lohika I'm not sure that we need to support so big "size_t" type. In most cases, we use "unit" for indexes.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl (hunk 3 above)
Comment (ozhydkov-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl:165]: resolve

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl (hunk 2 above)
Comment (ozhydkov-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/grid_sample_ref.cl]: resolve

**Comment by wilson-seok** [comment]
Please share conformance OpImplCheck test result also.

**Comment by ozhydkov-lohika** [comment]
@wilson-seok
This is conformance OpImplCheck test result :
Снимок экрана от 2022-10-26 13-20-26

**Review by wilson-seok — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/impls/ocl/grid_sample.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    using parent::parent;
+
+    std::unique_ptr<primitive_impl> clone() const override {
+        return make_unique<grid_sample_impl>(*this);
+    }
+
+    static primitive_impl* create(const grid_sample_node& arg, const kernel_impl_params& impl_param) {
+        auto params = get_default_params<kernel_selector::grid_sample_params>(impl_param);
+        auto optional_params =
+            get_default_optional_params<kernel_selector::grid_sample_optional_params>(arg.get_program());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/grid_sample.cpp:53]: arg.get_program() -> impl_param.prog

File: src/plugins/intel_gpu/src/graph/include/program_node.h [hunk 6]
```diff
@@ -536,7 +536,7 @@ template <class PType>
 struct typed_program_node : public typed_program_node_base<PType> {
     using typed_program_node_base<PType>::typed_program_node_base;
 
-    program_node& input() const { return program_node::get_dependency(0); }
+    program_node& input(size_t index = 0) const { return program_node::get_dependency(index); }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/program_node.h:539]: why not just use get_dependnecy(index) instead of fixing this?

**Comment by yeonbok** [comment]
LGTM if minor fixes are applied and CI passes.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/grid_sample/grid_sample_kernel_ref.cpp (hunk 1 above)
Comment (ozhydkov-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/grid_sample/grid_sample_kernel_ref.cpp]: resolved
add supported layout at GetSupportedKey

File: src/plugins/intel_gpu/src/graph/include/program_node.h (hunk 6 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/graph/include/program_node.h:539]: We have the same fix in this PR - https://github.com/openvinotoolkit/openvino/pull/13576. I just already made those changes for this operation.

**Review by wilson-seok — APPROVED** [review]

**Pull request merged**